rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,123 @@
+account created automatically without manual intervention.
+
+## Details
+
+The following are required for this feature:
+
+- Ability to automatically provision a Linux user if it's not present
+  on the node.
+- Ability to add the provisioned user to existing Linux groups defined
+  in the user traits/role.
+- Ability to add the provisioned user to sudoers.
+- Clean up the provisioned user / sudoers changes upon logout (being
+  careful not to remove pre-existing users).
+
+### Config/Role Changes
+
+Several new fields will need to be added to to the role `options` and
+`allow` sections:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  options:
+    # Controls whether this role supports auto provisioning of users.
+    auto_create_user: true
+    # Controls whether this role allows adding auto provisioned users to sudoers.
+    add_to_sudoers: true
+  allow:
+    # New field listing Linux groups to assign a provisioned user to.
+    # Should support user and identity provider traits like other fields (e.g. "logins")
+    groups: [ubuntu, "{{internal.groups}}", "{{external.xxx}}"]
+```
+
+An individual `ssh_service` can be configured disable auto user
+creation with the below config:
+
+```yaml
+ssh_service:
+    # when disabled, takes precedence over the role setting
+    disable_auto_create_user: true
+```
+
+### User creation
+
+A PAM module can be used in order to avoid running a Teleport process
+as root, and is already recommended in the [Teleport docs](https://goteleport.com/docs/server-access/guides/ssh-pam/#create-local-unix-users-on-login).
+
+This script will be adapted to support multiple Linux distributions
+and their differing methods by which users are created, and can be
+distributed as part of released Teleport packages.
+
+#### User Groups
+
+When a user is created they will be added to the specified groups from
+the role. In addition the user will be added to a special
+`teleport-tmp` group which can be used to indicate that the user was
+created by teleport and that its safe for it to be deleted.
+
+#### Valid user/group names
+
+The set of valid names that are valid on Linux varies between distros
+and are generally more restrictive than the allowed usernames in
+Teleport. This will require that names containing invalid characters
+have those characters removed/replaced. Information on the valid
+characters between Linux distros is available [here](https://systemd.io/USER_NAMES/).
+The common core of valid characters is `^[a-z][a-z0-9-]{0,30}$`.
+
+#### Adding and removing users from sudoers
+
+A user can be added to `sudoers` by creating files for each user in
+`/etc/sudoers.d/`.
+
+Each file will contain the following contents, allowing the specified
+user to execute root commands without a password:
+```
+${USERNAME} ALL=(ALL) ALL
+```
+
+##### Other considerations
+
+`sudo` isn't always available and [doas](https://man.openbsd.org/doas.conf.5)
```

Comment:
Is `doas` used on OpenBSD systems? We can leave it out of scope.

Reply:
Its primarily OpenBSD but it works on Linux,

Alpine linux is moving to consider `sudo` deprecated in the next release also.

> [Deprecation notes]
>    sudo will be moved to community in Alpine Linux 3.16. The suggested replacement is doas, which is available in main. See [TSC#1](https://gitlab.alpinelinux.org/alpine/tsc/-/issues/1)

https://alpinelinux.org/posts/Alpine-3.15.0-released.html